File: ~Programming_cheat_sheets/_scripting_languages/Python/data-structures-2020/__sorting/quick-sort/quick-sort.py
import array

class QuickSort:
    def __init__(self, arr):
        self.array = arr

    def sort(self):
        self.quickSort(0, len(arr)-1)

    def quickSort(self, left, right): # right is inclusive
        if left >= right: return
        p = self.partition(left, right)
        self.quickSort(left, p-1)
        self.quickSort(p+1, right)

    def partition(self, left, right): # right is inclusive
        wall = left
        pivot = self.array[right]
        for i in range(left, right+1):
            if self.array[i] < pivot:
                self.swap(i, wall)
                wall+=1
        self.swap(wall, right)
        return wall

    def swap(self, i, j):
        self.array[i], self.array[j] = self.array[j], self.array[i]


if __name__ == "__main__":
    arr = array.array('i', [5,4,6,3,7,2,8,1,0,9])
    quickSort = QuickSort(arr)
    quickSort.sort()
    print(arr)
    # array('i', [0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

File: ~other/~NEEDS_REVIEW/_compiled_languages/Java/_data-structures-and-algorithms/__sorting-algorithms/QuickSort.java
package com.example;

import java.util.Arrays;

class QuickSort {
    public static void sort(int[] array) {
        sort(array, 0, array.length - 1);
    }

    // recursive
    private static void sort(int[] array, int start, int end) {
        if (start >= end) return;
        int sortedPosition = partition(array, start, end);
        sort(array, start, sortedPosition - 1);
        sort(array, sortedPosition + 1, end);
    }

    private static int partition(int[] array, int low, int high) {
        int pivot = array[high];
        int i = low;

        // j will scan from left to right, looking for smaller then pivot elements
        // and swapping with i
        for (int j = low; j < high; j++) {
            // If current element is smaller than the pivot
            if (array[j] < pivot) {
                swap(array, i, j);
                i++; // move i into position, for future swaps!
            }
        }

        // swap arr[i] and arr[high] (or pivot)
        // because the element after i if the first of the second partition (elements higher than pivot)
        swap(array, i, high);

        // because i is now in the first element of the second partition (elements greater than the pivot)
        return i;
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}

public class QuickSortExample {
    public static void main(String[] args) {
        int[] array = new int[]{
                5, 6, 4, 7, 3, 8, 2, 9, 1, 0
        };
        QuickSort.sort(array);
        System.out.println(Arrays.toString(array));
        //=> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
    }
}
